### DISCUSSION TITLE
Communication Issues

### DISCUSSION THREAD

**Discussion by sfritz85** [Dilbert66/esphome-dsckeybus#76] [General]
Having some problems getting this to work reliably.  Using Nodemcu 8266 on a dsc PC5020 with 4 partitions.  The NodeMCU appears to keep resetting randomly.  Any suggestions greatly appreciated.

sample log:

INFO Reading configuration /config/esphome/DscAlarm.yaml...
INFO Starting log output from dscalarm.local using esphome API
INFO Successfully connected to dscalarm.local
[11:30:05][I][app:102]: ESPHome version 2022.12.8 compiled on Feb  7 2023, 10:08:23
[11:30:05][C][status_led:019]: Status LED:
....
[11:31:16][D][text_sensor:067]: 'dscalarm System Status': Sending state 'online'
[11:31:50][D][text_sensor:067]: 'dscalarm Partition 4 Status ': Sending state 'disarmed'
[11:31:57][D][text_sensor:067]: 'dscalarm Partition 1 Status ': Sending state 'disarmed'
[11:31:58][D][text_sensor:067]: 'dscalarm Partition 3 Status ': Sending state 'armed_away'
[11:32:06][D][text_sensor:067]: 'dscalarm Partition 2 Status ': Sending state 'disarmed'
[11:32:06][D][text_sensor:067]: 'dscalarm Partition 1 Msg ': Sending state '01: Ready'
[11:32:07][D][text_sensor:067]: 'dscalarm Partition 4 Msg ': Sending state '02: Stay zones open'
[11:32:07][D][text_sensor:067]: 'dscalarm Partition 2 Msg ': Sending state '01: Ready'
[11:32:11][D][text_sensor:067]: 'dscalarm Partition 3 Msg ': Sending state '05: Armed away'
[11:32:16][D][text_sensor:067]: 'dscalarm System Status': Sending state 'online'
[11:32:50][D][Debug22:247]: Panel command data: A5,00,23,08,EB,80,00,00,3B,00
[11:32:50][D][Debug33:254]: Partition data 00: 91,01,00,00,00,00,00,00,00,00,01,00,00
[12:18:51][D][text_sensor:067]: 'dscalarm Partition 1 Status ': Sending state 'disarmed'
INFO dscalarm.local: Error while reading incoming messages: Error while reading data: 0 bytes read on a total of 3 expected bytes
INFO dscalarm.local: Ping Failed: Error while reading data: 0 bytes read on a total of 3 expected bytes
INFO Disconnected from ESPHome API for dscalarm.local
WARNING Disconnected from API
INFO Successfully connected to dscalarm.local
[12:19:10][D][Debug22:247]: Panel command data: 05,00,91,02,91,01,92,05,91,02
[12:19:10][D][Debug33:254]: Partition data 00: 91,02,00,00,00,00,00,00,00,00,01,00,00
[12:19:12][D][Debug22:247]: Panel command data: 27,00,91,02,91,01,00,4C,00,00
[12:19:12][D][Debug22:247]: Panel command data: 05,00,91,01,91,01,92,05,91,02
[12:19:12][D][Debug33:254]: Partition data 00: 91,01,00,00,00,00,00,00,00,00,01,00,00
[12:19:12][D][Debug33:254]: Partition data 02: 92,05,01,01,00,00,00,00,00,00,00,00,00

[12:24:18][D][Debug22:247]: Panel command data: 27,00,91,02,91,01,00,4C,00,00
[12:24:18][D][Debug33:254]: Partition data 00: 91,02,00,00,00,00,00,00,00,00,01,00,00
[12:24:18][D][Debug22:247]: Panel command data: 05,00,91,01,91,01,92,05,91,02
[12:24:18][D][Debug33:254]: Partition data 00: 91,01,00,00,00,00,00,00,00,00,01,00,00
[12:24:18][D][Debug33:254]: Partition data 03: 91,02,00,00,00,00,00,00,00,00,01,00,00
[12:24:20][D][text_sensor:067]: 'dscalarm Partition 1 Msg ': Sending state '01: Ready'
[12:24:26][D][text_sensor:067]: 'dscalarm Partition 3 Msg ': Sending state '05: Armed away'
[12:24:26][D][api:102]: Accepted 192.168.18.251
[12:24:26][D][api.connection:918]: Home Assistant 2023.2.2 (192.168.18.251): Connected successfully
INFO dscalarm.local: Ping timed out!
INFO Disconnected from ESPHome API for dscalarm.local
WARNING Disconnected from API
INFO Successfully connected to dscalarm.local
[12:24:39][D][api:102]: Accepted 192.168.18.251
[12:24:40][D][api.connection:918]: Home Assistant 2023.2.2 (192.168.18.251): Connected successfully
INFO dscalarm.local: Error while reading incoming messages: Error while reading data: [Errno 104] Connection reset by peer
INFO dscalarm.local: Ping Failed: Error while reading data: [Errno 104] Connection reset by peer
INFO Disconnected from ESPHome API for dscalarm.local
WARNING Disconnected from API
INFO Successfully connected to dscalarm.local

**Comment by Dilbert66** [comment]
Hard to say what the issue is.  Serial logs would be more helpful. What I do recommend is that you try the "dev" branch first as that one has many enhancements and bug fixes.

**Reply by sfritz85** [reply]
Is serial logs when connecting nodemcu to serial port then view logs?

How do i access dev branch?

**Comment by Dilbert66** [comment]
yes, that's when you connect to the serial port and use a serial terminal to view the logs.
You select the branch from the top left of the page. In the drop down select "dev" instead of "master".
Heres a direct link:
https://github.com/Dilbert66/esphome-dsckeybus/tree/dev

**Reply by sfritz85** [reply]
Thanks for your quick response, i will try that.

**Comment by sfritz85** [comment]
Dev update appears to be working more reliably now but still has some dropouts (maybe bad wifi connection?).  I did find other questions/issues:
1) defaultPartition set to 2 in my instance.  When i try to arm alarm from HA i get the error: Failed to call service.alarm_control_panel/alarm_arm_away.  extra keys not allowed @ data['partition']
2). DSC system PC5020 shows a communication error.
3).  Is it required to uncomment zone expander addresses for all 43 active zones?
[logs_DscAlarm_logs_comm.txt](https://github.com/Dilbert66/esphome-dsckeybus/files/10700310/logs_DscAlarm_logs_comm.txt)
[logs_DscAlarm_logs (8).txt](https://github.com/Dilbert66/esphome-dsckeybus/files/10700313/logs_DscAlarm_logs.8.txt)

**Comment by sfritz85** [comment]
I tried updating to latest version of dev Mar, 7/23.  Unfortunately i still get this message. 
  Error executing script. Invalid data for call_service at pos 1: extra keys not allowed @ data['partition']
  
 Any  help greatly appreciated.  Thanks in advance.

**Reply by Dilbert66** [reply]
There is no "partition" parameter for service alarm_arm_away. It will arm whichever is set as the default partition.
Leave the 2 zone expander  entries at 0 for as they are not needed for most people.  Those entries are for expander device addresses, not zones.  The COM (communications) and ZF (zone fault) errors , I have no idea why you are getting those.  Possibly you have a zone expander address set that conflicts with something on your system.  Set them to 0 to ensure that does not happen.

**Comment by sfritz85** [comment]
I have tried arming both away and home.  Same error.  Set default partition to 2.  Also need to arm/disarm other partitions 1, 3 and 4 as well separately.
I have the following setup in yaml:
 defaultPartition: "2" 

#zone expander addresses:
#9  - zones 9-16
#10 - zones 17-24
#11 - zones 25-32
#12 - zones 33-40 (for systems with 64 zone support)
#13 - zones 41-48 (for systems with 64 zone support)
#14 - zones 49-56 (for systems with 64 zone support)
#16 - zones 57-64 (for systems with 64 zone support)  
  expanderAddr1: "0" # 1st zone expander emulator address to use . Set to 0 to disable. 
  expanderAddr2: "0" # 2nd expander emulator address to use . Set to 0 to disable.

**Comment by Dilbert66** [comment]
Like I said, the error you show indicates that you are sending extra info the arm_away and arm_home commands.  They do not have parameters.
Alternatively you can use two other ways to send commands to the panel.  You can use the "set_alarm_state("state","code","partition") where "state" is A for away, S for stay/home N for night. and code is the arm code if needed. You can send an empty string.  Partition is the partiton to arm: 1, 2,3, etc
The other way is to send direct commands to the panel using the service alarm_keypress_partition("keys","partition") where keys is the key sequence.  For arm way it's "A", etc. Like the State command. You can also send bypass commands using that same service.  For example to bypass all open zones, send "*199#", or to bypass zone 5, send *105#

**Reply by sfritz85** [reply]
Thanks, that helped a lot.

**Comment by sfritz85** [comment]
Is it possible to have one physical alarm panel to control 4 separate partitions as separate alarm panels in HA? I am trying to setup 1 alarm panel with 4 partitions with partition 2 is default.  However with the code below i only get default (partition 2) alarm panel to select from.  Is there a way to have 4 independent alarm panels to select from?
Attached is code i tried unsuccessfully.  Thanks in advance for your help with this.
[alarm.txt](https://github.com/Dilbert66/esphome-dsckeybus/files/10944300/alarm.txt)

**Comment by sfritz85** [comment]
Recommendations for above setup greatly appreciated.

**Comment by sfritz85** [comment]
Compiled with latest ESPHome 2023.3.1.  I get the following messages:

Compiling /data/dscalarm/.pioenvs/dscalarm/src/dscKeybusInterface.cpp.o
src/dscKeybusInterface.cpp: In member function 'bool dscKeybusInterface::validCRC()':
src/dscKeybusInterface.cpp:511:3: warning: this 'if' clause does not guard... [-Wmisleading-indentation]
  511 |   if (dataSum % 256 == panelData[byteCount])
      |   ^~
src/dscKeybusInterface.cpp:514:7: note: ...this statement, but the latter is misleadingly indented as if it were guarded by the 'if'
  514 |       return false;
      |       ^~~~~~
src/dscKeybusInterface.cpp: In static member function 'static void dscKeybusInterface::dscClockInterrupt()':
src/dscKeybusInterface.cpp:583:69: warning: 'static bool dscKeybusInterface::redundantPanelData(byte*, volatile byte*, byte)' is deprecated: Use IRAM_ATTR in place of ICACHE_RAM_ATTR to move functions into IRAM [-Wdeprecated-declarations]
  583 |         if (redundantPanelData(pcmd, isrPanelData, isrPanelByteCount)) {
      |                                                                     ^
src/dscKeybusInterface.cpp:495:1: note: declared here
  495 | dscKeybusInterface::redundantPanelData(byte  previousCmd[] , volatile byte  currentCmd[], byte checkedBytes) {
      | ^~~~~~~~~~~~~~~~~~
src/dscKeybusInterface.cpp: In static member function 'static void dscKeybusInterface::processPendingResponses(byte)':
src/dscKeybusInterface.cpp:815:62: warning: 'static void dscKeybusInterface::updateWriteBuffer(byte*, int, byte, int, bool)' is deprecated: Use IRAM_ATTR in place of ICACHE_RAM_ATTR to move functions into IRAM [-Wdeprecated-declarations]
  815 |     updateWriteBuffer((byte * ) moduleSlots, 9,1,maxFields11 );
      |                                                              ^
src/dscKeybusInterface.cpp:785:1: note: declared here
  785 | dscKeybusInterface::dscKeybusInterface::updateWriteBuffer(byte * src, int bit,byte partition,int len,  bool alarm) {
      | ^~~~~~~~~~~~~~~~~~
src/dscKeybusInterface.cpp:836:53: warning: 'static void dscKeybusInterface::updateWriteBuffer(byte*, int, byte, int, bool)' is deprecated: Use IRAM_ATTR in place of ICACHE_RAM_ATTR to move functions into IRAM [-Wdeprecated-declarations]
  836 |         updateWriteBuffer((byte *) cmdD0buffer,9,1,6);
      |                                                     ^
src/dscKeybusInterface.cpp:785:1: note: declared here
  785 | dscKeybusInterface::dscKeybusInterface::updateWriteBuffer(byte * src, int bit,byte partition,int len,  bool alarm) {
      | ^~~~~~~~~~~~~~~~~~
src/dscKeybusInterface.cpp: In static member function 'static void dscKeybusInterface::processPendingQueue(byte)':
src/dscKeybusInterface.cpp:857:164: warning: 'static void dscKeybusInterface::updateWriteBuffer(byte*, int, byte, int, bool)' is deprecated: Use IRAM_ATTR in place of ICACHE_RAM_ATTR to move functions into IRAM [-Wdeprecated-declarations]
  857 |     updateWriteBuffer((byte * ) writeQueue[outIdx].data,  writeQueue[outIdx].writeBit,writeQueue[outIdx].partition,writeQueue[outIdx].len, writeQueue[outIdx].alarm); //populate write buffer and set ready to send flag
      |                                                                                                                                                                    ^
src/dscKeybusInterface.cpp:785:1: note: declared here
  785 | dscKeybusInterface::dscKeybusInterface::updateWriteBuffer(byte * src, int bit,byte partition,int len,  bool alarm) {
      | ^~~~~~~~~~~~~~~~~~
Compiling /data/dscalarm/.pioenvs/dscalarm/src/dscKeybusPrintData.cpp.o

**Reply by Dilbert66** [reply]
Ignore. Just warnings

**Comment by sfritz85** [comment]
How can i find out user who last armed / disarmed  a partition?

**Comment by Dilbert66** [comment]
View events using keypad.  Events are also sent to the event text sensor
